### PR TITLE
Create benchmark for batched time_delay_estimator based on synthetic data #44

### PR DESCRIPTION
**Pull request by GFariasR**

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -679,6 +681,14 @@ jb_fftw_bm_time_delay_estimator_LDADD = \
 jb_fftw_bm_time_delay_estimator_LDFLAGS = \
   $(FFTW3_LDFLAGS)
 
+jb_fftw_bm_time_delay_estimator_many_SOURCES = jb/fftw/bm_time_delay_estimator_many.cpp
+jb_fftw_bm_time_delay_estimator_many_CPPFLAGS =
+jb_fftw_bm_time_delay_estimator_many_LDADD = \
```
Comment (coryan) [inline Makefile.am:698]: Please notice that I separated *_LDFLAGS from *_LDADD.  You need to add:

```
jb_fftw_bm_time_delay_estimator_many_LDFLAGS = $(FFTW3_LDFLAGS)
```

File: jb/fftw/time_delay_estimator_many.hpp [hunk 2]
```diff
@@ -108,7 +112,8 @@ class time_delay_estimator_many {
     // Validate the input sizes.  For some types of timeseries the
     // alignment may be different too, but we only use the alignment
     // when the type of timeseries guarantees to always be aligned.
-    if (a.size() != tmpa_.size() or b.size() != tmpa_.size()) {
+    if (jb::detail::element_count(a) != jb::detail::element_count(tmpa_) or
+        jb::detail::element_count(b) != jb::detail::element_count(tmpa_)) {
```
Comment (coryan) [inline jb/fftw/time_delay_estimator_many.hpp]: s/tmpa_/tmpb_/ ??

File: jb/fftw/time_delay_estimator_many.hpp [hunk 3]
```diff
@@ -108,7 +112,8 @@ class time_delay_estimator_many {
     // Validate the input sizes.  For some types of timeseries the
     // alignment may be different too, but we only use the alignment
     // when the type of timeseries guarantees to always be aligned.
-    if (a.size() != tmpa_.size() or b.size() != tmpa_.size()) {
+    if (jb::detail::element_count(a) != jb::detail::element_count(tmpa_) or
```
Comment (coryan) [inline jb/fftw/time_delay_estimator_many.hpp]: You request the same **shape** in the constructor, and just the same element count here.  I think you want the same shape in both cases.

File: jb/fftw/time_delay_estimator_many.hpp [hunk 4]
```diff
@@ -128,18 +133,17 @@ class time_delay_estimator_many {
...
-      for (std::size_t j = 0; j != nsamples_; ++j, ++k, ++it_out) {
+      for (std::size_t j = 0; j != nsamples_; ++j, ++it_out) {
         if (max_val < *it_out) {
           max_val = *it_out;
           tde_val = j;
         }
       }
-      if (jb::testing::check_close_enough(sum2[i], precision_type{0}, 1)) {
+      if (sum2[i] < std::numeric_limits<precision_type>::epsilon()) {
         confidence[i] = std::numeric_limits<precision_type>::max();
```
Comment (coryan) [inline jb/fftw/time_delay_estimator_many.hpp]: I did not catch this earlier, but do we really want confidence to be max when basically we have 0 information?  I would think confidence == 0 would be more appropriate.

File: jb/fftw/ut_time_delay_estimator_many.cpp [hunk 5]
```diff
@@ -1,8 +1,10 @@
 #include <jb/fftw/tde_result.hpp>
```
Comment (coryan) [inline jb/fftw/ut_time_delay_estimator_many.cpp]: Having this include in the test masked the problem with the missing tde_result.hpp in time_delay_estimator_many.hpp.

This is why we try to include the main header first, to root out missing headers earlier.

File: jb/fftw/ut_time_delay_estimator_many.cpp [hunk 6]
```diff
@@ -17,14 +19,15 @@ BOOST_AUTO_TEST_CASE(fftw_time_delay_estimator_many_3_dim_tde_with_0) {
   int const nsamples = 1 << 15;
   int const S = 20;
   int const V = 4;
-  int const argmax_tol = 2;
+  int const argmax_tol = 3;
```
Comment (coryan) [inline jb/fftw/ut_time_delay_estimator_many.cpp]: Why?  Please add a comment, or justify the number somehow.

File: jb/fftw/ut_time_delay_estimator_many.cpp [hunk 7]
```diff
@@ -81,14 +83,15 @@ BOOST_AUTO_TEST_CASE(fftw_time_delay_estimator_many_3_dim_tde_delay_0) {
...
```
Comment (coryan) [inline jb/fftw/ut_time_delay_estimator_many.cpp]: Why?

File: jb/fftw/ut_time_delay_estimator_many.cpp [hunk 8]
```diff
@@ -1178,3 +1180,35 @@ BOOST_AUTO_TEST_CASE(fftw_time_delay_estimator_many_vector_tde_complex_double) {
           confidence, expected_confidence, confidence_tol),
       "confidence is not within tolerance(" << confidence_tol << ")");
 }
+
+/**
+ * @test Verify thattime delay estimator handles errors
+ */
+BOOST_AUTO_TEST_CASE(fftw_time_delay_estimator_many_errors) {
```
Comment (coryan) [inline jb/fftw/ut_time_delay_estimator_many.cpp]: This PR is about two things: improving test coverage, and definiing the benchmark.  Can send two separate PRs?  As things stand the test fixes will be queued until we agree on what the benchmark is supposed to do.

File: jb/testing/microbenchmark_config.cpp [hunk 9]
```diff
@@ -38,6 +38,10 @@ namespace defaults {
 #define JB_DEFAULTS_microbenchmark_size 0
 #endif
 
+#ifndef JB_DEFAULTS_microbenchmark_timeseries
```
Comment (coryan) [inline jb/testing/microbenchmark_config.cpp]: Please create a new configuration class in bm_time_delay_estimator_many.  This is supposed to be the configuration class for all microbenchmarks, that "timeseries" parameter would not apply to many of them.  I was just lucky in bm_time_delay_estimator.cpp because no new configuration parameters were needed.

You can look at jb/itch5/bm_order_book.cpp for an example.

File: jb/testing/tde_square_sum.hpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+/**
+ * Compute the sum square of a family of timeseries.
+ *
+ * @return tde_result with the square sum of each timeseries
+ *
+ * @tparam container_t type of the collection of timeseries
+ */
+template <typename container_t>
+auto sum_square(container_t const& ts) {
```
Comment (coryan) [inline jb/testing/tde_square_sum.hpp]: The function name and file name do not match, please fix one or the other.

File: jb/fftw/bm_time_delay_estimator_many.cpp [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+int main(int argc, char* argv[]) try {
+  jb::testing::microbenchmark_config cfg;
+  cfg.test_case("float:aligned").process_cmdline(argc, argv);
+
+  std::cout << "Configuration for test\n" << cfg << std::endl;
+
+  // check if it is 2 dimensional
+  bool dim_2 = cfg.timeseries() > 1;
+
+  if (cfg.test_case() == "float:aligned") {
```
Comment (coryan) [inline jb/fftw/bm_time_delay_estimator_many.cpp]: What we need to do is compare the performance of 100 calls to time_delay_estimator<> vs. 1 call to time_delay_estimator_many<> with an array of 100 elements.  So I would do something like this:

```
  if (cfg.test_case() == "float:aligned:many") {
    benchmark_many<jb::fftw::aligned_multi_array<float, 2>>(cfg);
  } else if (cfg.test_case() == "float::aligned::single") {
    benchmark_single<jb::fftw::aligned_vector<float>>(cfg);
  }
```

Then you modify the `fixture::run()` to call `estimate` 100 times for vectors, and 1 time for an array of 100 rows.

Obviously the 100 value is a parameter.

**Review by coryan — CHANGES_REQUESTED** [review]
Please split this in two PRs.  I left comments on how to structure the benchmark.

File: jb/fftw/time_delay_estimator_many.hpp (hunk 4 above)
Comment (GFariasR) [reply jb/fftw/time_delay_estimator_many.hpp]: ACK.

I see, I thought that sqr sum of the elements closer to 0 will compute a confidence = X / very small number => a very large number independently of X. Anyway if sum2 ~= 0 means: we do not know! 

I will fix it.

File: jb/fftw/time_delay_estimator_many.hpp (hunk 3 above)
Comment (GFariasR) [reply jb/fftw/time_delay_estimator_many.hpp]: ACK. I saw you cherry-picked and fixed. We can close this one

File: jb/fftw/time_delay_estimator_many.hpp (hunk 2 above)
Comment (GFariasR) [reply jb/fftw/time_delay_estimator_many.hpp]: ACK. Saw you fixed. Close.

File: jb/fftw/ut_time_delay_estimator_many.cpp (hunk 5 above)
Comment (GFariasR) [reply jb/fftw/ut_time_delay_estimator_many.cpp]: ACK. Got it. Removed from _ut.cpp.

File: jb/fftw/ut_time_delay_estimator_many.cpp (hunk 6 above)
Comment (GFariasR) [reply jb/fftw/ut_time_delay_estimator_many.cpp]: ACK. I do not know what you did with the library but now tolerance = 1 tests are passing but for complex<float> which make sense because of the argument of a complex number is yet another computation sqrt(a^2 - b^2) if my memory does not fail here. 
Anyway, tolerances are back to 1 except for the aforementioned combination, where a comment was added.

File: jb/fftw/ut_time_delay_estimator_many.cpp (hunk 7 above)
Comment (GFariasR) [reply jb/fftw/ut_time_delay_estimator_many.cpp]: ACK. Fixed

File: jb/fftw/ut_time_delay_estimator_many.cpp (hunk 8 above)
Comment (GFariasR) [reply jb/fftw/ut_time_delay_estimator_many.cpp]: ACK. What do you want me to do now that you fixed the testing part. Should I continue pushing the existing one?

File: jb/testing/tde_square_sum.hpp (hunk 10 above)
Comment (GFariasR) [reply jb/testing/tde_square_sum.hpp]: ACK, I think sum_square explains what is intended, so I will change the file name

File: jb/fftw/bm_time_delay_estimator_many.cpp (hunk 11 above)
Comment (GFariasR) [reply jb/fftw/bm_time_delay_estimator_many.cpp]: ACK. I will work on this tomorrow. Now I am going to push the changes to my master to see if everything is fine. There are 2 issues still pending, this one and the benchmark config.
I will fix them tomorrow if you prefer to wait for the whole push

File: jb/testing/microbenchmark_config.cpp (hunk 9 above)
Comment (GFariasR) [reply jb/testing/microbenchmark_config.cpp]: ACK. Working on it now.

File: Makefile.am (hunk 1 above)
Comment (GFariasR) [reply Makefile.am:698]: ACK, done

File: .gitignore [hunk 12]
```diff
@@ -46,7 +46,7 @@ tsan/
 asan/
 ubsan/
 opt/
-build/
+build*/
```
Comment (coryan) [inline .gitignore]: This is fine by me, but I am curious, what are you calling your build directories?

File: jb/fftw/bm_time_delay_estimator_many.cpp [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+#include <jb/log.hpp>
+
+#include <chrono>
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+/**
+ * This benchmark is implemented to evaluate the performance of
```
Comment (coryan) [inline jb/fftw/bm_time_delay_estimator_many.cpp]: Consider:

```
 * This benchmark evaluates the performance of time_delay_estimator_many.
 *
 * It validates the assumption that FFTW performs better with an array of timeseries vs.
 * being called multiple times blah blah...
```

File: jb/fftw/bm_time_delay_estimator_many.cpp [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+ * better if we pass a multi array container with n time series in one call,
+ * than having
+ * to call the library n times passing a container (like vector) with one time
+ * series each
+ * time
+ */
+
+namespace {
+
+// We expect delays to be around 1250 microseconds, max delays are
```
Comment (coryan) [inline jb/fftw/bm_time_delay_estimator_many.cpp]: About that comment, it was true when I wrote it ages ago.... consider instead:

```
// These magic numbers are motivated by observed delays between two market
// feeds.  They assume that delay is normally around 1,250 microseconds, but
// can be as large as 6,000 microseconds.  To reliably detect the 6,000 usecs
// delays we need samples that cover at least 18,000 microseconds, assuming
// a sampling rate of 10 microseconds that is 1,800 samples, FFTs work well
// with sample sizes that are powers of 2, so we use 4096 samples.
std::chrono::microseconds const expected_delay(1250)
std::chrono::microseconds const sampling_period(10);
int const nsamples = 4096;
```

File: jb/fftw/bm_time_delay_estimator_many.cpp [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+                  " to compute TDE passing a container with one time series as "
+                  "argument every time. "
+                  "If it is *:many, the fixture uses a 2-dimension array "
+                  "containing this "
+                  "many time series as argument to a one time compute TDE."),
+          this, defaults::n_timeseries) {
+}
+
+void config::validate() const {
+  log().validate();
```
Comment (coryan) [inline jb/fftw/bm_time_delay_estimator_many.cpp:297]: You should probably call `microbenchmark().validate();` even if it is a no-op.

Comment (coryan) [reply jb/fftw/bm_time_delay_estimator_many.cpp:297]: And I should too!  I am fixing `jb/itch5/bm_order_book.cpp`

**Review by coryan — CHANGES_REQUESTED** [review]
Looks really good.  I cannot find any problem in your code that would explain why you are having problems setting the `microbenchmark.test-case` command-line argument.

I would check:
* Do a clean build, just in case.
* Print out the preprocessed output of `jb/fftw/bm_time_delay_estimator_many.cpp` (that is a combination of make V=1 and gcc -E).

If that fails, just address the changes I suggested we can commit and I can try to debug in my machine.

File: .gitignore (hunk 12 above)
Comment (GFariasR) [reply .gitignore]: ACK.
buildUbuntu14.04
buildUbuntu16.10
buildFed.... you got the idea, right?

File: jb/fftw/bm_time_delay_estimator_many.cpp (hunk 13 above)
Comment (GFariasR) [reply jb/fftw/bm_time_delay_estimator_many.cpp]: ACK.

Working on it.

File: .gitignore (hunk 12 above)
Comment (coryan) [reply .gitignore]: Ack.  You know you can build inside `build/Ubuntu16.04`? (notice the extra / ).

Just do:

```
mkdir -p build/Ubuntu16.04
cd build/Ubuntu16.04
../../configure 
```

File: jb/fftw/bm_time_delay_estimator_many.cpp (hunk 14 above)
Comment (GFariasR) [reply jb/fftw/bm_time_delay_estimator_many.cpp]: ACK. Updated

File: jb/fftw/bm_time_delay_estimator_many.cpp (hunk 15 above)
Comment (GFariasR) [reply jb/fftw/bm_time_delay_estimator_many.cpp:297]: ACK. Done

File: .gitignore (hunk 12 above)
Comment (GFariasR) [reply .gitignore]: ACK. I will implement the change right now. I am reverting that commit...

**Review by coryan — APPROVED** [review]
